### PR TITLE
[SELS-TASK] Integrate follow on profile page

### PR DESCRIPTION
**Pull request by jstephend-sun**
Subbranch of PR 32

Asana [Task](https://app.asana.com/0/1201478050789792/1201633725750030/f)

expected output:
can navigate to
/users/1 or users/{id}
to show the profile page of the user

the follow button is functional
can click follow to follow user, then it will change to unfollow button
can click unfollow button to unfollow

the number of followers will also change 

preconditions:
user must be logged in

commands
```
cd frontend
npm i
npm start
```
```
cd backend
composer install
php artisan migrate:fresh --seed
php artisan serve
```

screenshots
image

image

### CODE REVIEW

File: frontend/src/components/UserProfile/UserProfile.tsx [hunk 1]
```diff
@@ -34,23 +34,61 @@ interface Props {
...
+        setIsFollowing(false);
+        fetchUserWithFollows(cookies.token, id);
+      });
+    } else {
+      followUser(user?.id, cookies.token, () => {
+        setLoading(false);
+        setIsFollowing(true);
+        fetchUserWithFollows(cookies.token, id);
+      });
+    }
```
Comment (ejanton-sun) [inline frontend/src/components/UserProfile/UserProfile.tsx]: This looks redundant, please refactor. If you have problems regarding on this, please don't hesitate to ask me.

**Review by ejanton-sun — CHANGES_REQUESTED** [review]

File: frontend/src/components/UserProfile/UserProfile.tsx (hunk 1 above)
Comment (jstephend-sun) [reply frontend/src/components/UserProfile/UserProfile.tsx]: done po

File: frontend/src/components/UserProfile/UserProfile.tsx [hunk 2]
```diff
@@ -64,19 +64,14 @@ const _UserProfile = ({
...
-      });
-    }
+    let callback = () => {
+      setLoading(false);
+      setIsFollowing(false);
+      fetchUserWithFollows(cookies.token, id);
+    };
+    isFollowing
+      ? unfollowUser(user?.id, cookies.token, callback)
+      : followUser(user?.id, cookies.token, callback);
```
Comment (ejanton-sun) [inline frontend/src/components/UserProfile/UserProfile.tsx]: This is good for working, there is another way to simplify this. This topic is related to [destructuring in js](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment)
```js
let data = {
   user_id: user?.id,
   cookies_token: cookies.token,
   callback: () => {
      setLoading(false);
      setIsFollowing(false);
      fetchUserWithFollows(cookies.token, id);
   }
} 

isFollowing ? unfollowUser(data) : followUser(data);
```
And after this, update your `followUser` and `unfollowUser` method that accepts on parameter.
```js
followUser({ user_id, cookies_token, callback}){
//code
}
unfollowUser({ user_id, cookies_token, callback}){
//code
}
```

**Review by ejanton-sun — CHANGES_REQUESTED** [review]
It is better to accept one argument so that every time you add new column/field the code will have minimal changes.

File: frontend/src/actions/users.ts [hunk 3]
```diff
@@ -70,17 +70,20 @@ export const fetchUserWithFollows = (
...
 
-export const followUser = (
-  userId: number,
-  token: string,
-  callback: Function = () => {}
-) => {
+export const followUser = (data: {
+  user_id: number;
+  token: string;
+  callback: Function;
```
Comment (ejanton-sun) [inline frontend/src/actions/users.ts:76]: this will result and error because `;` instead of `,`

File: frontend/src/actions/users.ts [hunk 4]
```diff
@@ -64,13 +70,20 @@ export const fetchUserWithFollows = (token: string, id?: number | string) => {
...
-  return (dispatch: Dispatch) => {
+export const followUser = (data: {
+  user_id: number;
+  token: string;
+  callback: Function;
+}) => {
+  return async (dispatch: Dispatch) => {
     backend.get('/sanctum/csrf-cookie').then(async (csrf_response) => {
+      let user_id, token, callback;
+      ({ user_id, token, callback } = data);
```
Comment (ejanton-sun) [inline frontend/src/actions/users.ts]: Can be done in one line
```suggestion
      let { user_id, token, callback } = data;
```

File: frontend/src/actions/users.ts [hunk 5]
```diff
@@ -79,22 +92,31 @@ export const followUser = (userId: number, token: string) => {
...
     });
   };
 };
 
-export const unfollowUser = (userId: number, token: string) => {
-  return (dispatch: Dispatch) => {
+export const unfollowUser = (data: {
+  user_id: number;
+  token: string;
+  callback: Function;
```
Comment (ejanton-sun) [inline frontend/src/actions/users.ts:107]: this will result an error same as above

File: frontend/src/actions/users.ts [hunk 6]
```diff
@@ -79,22 +92,31 @@ export const followUser = (userId: number, token: string) => {
...
+}) => {
+  return async (dispatch: Dispatch) => {
     backend.get('/sanctum/csrf-cookie').then(async (csrf_response) => {
+      let user_id, token, callback;
+      ({ user_id, token, callback } = data);
```
Comment (ejanton-sun) [inline frontend/src/actions/users.ts]: Can be done in single line
```suggestion
      let { user_id, token, callback } = data;
```

File: frontend/src/components/UserList/UserList.tsx [hunk 7]
```diff
@@ -37,13 +37,23 @@ export const UserList = ({
...
+      token: cookies.token,
+      callback: callback,
+    };
     if (params.row.is_following) {
-      unfollowUser(rowId, cookies.token);
+      unfollowUser(data);
     } else {
-      followUser(rowId, cookies.token);
+      followUser(data);
     }
```
Comment (ejanton-sun) [inline frontend/src/components/UserList/UserList.tsx]: use ternary
```js
params.row.is_following ? unfollowUser(data) : followUser(data);
```

**Review by ejanton-sun — CHANGES_REQUESTED** [review]

File: frontend/src/actions/users.ts (hunk 3 above)
Comment (jstephend-sun) [reply frontend/src/actions/users.ts:76]: no error

File: frontend/src/actions/users.ts (hunk 4 above)
Comment (jstephend-sun) [reply frontend/src/actions/users.ts]: done

File: frontend/src/actions/users.ts (hunk 5 above)
Comment (jstephend-sun) [reply frontend/src/actions/users.ts:107]: no error

File: frontend/src/actions/users.ts (hunk 6 above)
Comment (jstephend-sun) [reply frontend/src/actions/users.ts]: done

File: frontend/src/components/UserList/UserList.tsx (hunk 7 above)
Comment (jstephend-sun) [reply frontend/src/components/UserList/UserList.tsx]: done

**Review by ejanton-sun — APPROVED** [review]

**Pull request merged**